### PR TITLE
 BUG guess_Datetime_format doesn't guess 27.03.2003 14:55:00.000

### LINKED ISSUE #50317: BUG guess_Datetime_format doesn't guess 27.03.2003 14:55:00.000
```python
----> 1 guess_datetime_format('27.03.2003 14:55:00.000')

File ~/pandas-dev/pandas/_libs/tslibs/parsing.pyx:967, in pandas._libs.tslibs.parsing.guess_datetime_format()
    965 print('tokens[i]', tokens[i])
    966 print('padding', padding)
--> 967 token_filled = _fill_token(tokens[i], padding)
    968 if token_format is None and token_filled == parsed_formatted:
    969     format_guess[i] = attr_format

File ~/pandas-dev/pandas/_libs/tslibs/parsing.pyx:1026, in pandas._libs.tslibs.parsing._fill_token()
   1024 print('token', token)
   1025 seconds, nanoseconds = token.split(".")
-> 1026 seconds = f"{int(seconds):02d}"
   1027 # right-pad so we get nanoseconds, then only take
   1028 # first 6 digits (microseconds) as stdlib datetime

ValueError: invalid literal for int() with base 10: ''
```

It should return `None`, or guess the format, but not error

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #50317 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/parsing.pyx [hunk 1]
```diff
@@ -1016,9 +1016,11 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
...
 cdef str _fill_token(token: str, padding: int):
     cdef str token_filled
-    if "." not in token:
+    if re.search(r"\d*\.\d+", token) is None:
+        # For example: 98
         token_filled = token.zfill(padding)
     else:
+        # For example: 00.123
         seconds, nanoseconds = token.split(".")
         seconds = f"{int(seconds):02d}"
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/parsing.pyx:1025]: previously,
```
seconds = f"{int(seconds):02d}"
```
was failing if `token` was `.`

If we make the check more precise, then we avoid that issue

**Review by mroeschke — COMMENTED** [review]
LGTM. Doesn't need a whatsnew right?

File: pandas/_libs/tslibs/parsing.pyx [hunk 2]
```diff
@@ -1016,9 +1016,11 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
 
 cdef str _fill_token(token: str, padding: int):
     cdef str token_filled
-    if "." not in token:
+    if re.search(r"\d*\.\d+", token) is None:
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/parsing.pyx:1019]: is the leading `\d*` needed here?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/parsing.pyx]: `\d+` would be better actually, thanks

This is for when it corresponds with the `%S.%f` directive

**Comment by MarcoGorelli** [comment]
> LGTM. Doesn't need a whatsnew right?

thanks! and no, this wouldn't (yet) make a user-facing difference

it fixes https://github.com/pandas-dev/pandas/issues/2586 by accident, but I'd say that that fits into the PDEP4-related part of the whatsnew (as in, PDEP4 should have fixed that, it just didn't quite because of an error in implementation here)

**Review by mroeschke — APPROVED** [review]
LGTM. Merge when ready @jbrockmendel

File: pandas/_libs/tslibs/parsing.pyx [hunk 3]
```diff
@@ -1016,9 +1016,11 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
 
 cdef str _fill_token(token: str, padding: int):
     cdef str token_filled
-    if "." not in token:
+    if re.search(r"\d+\.\d+", token) is None:
+        # For example: 98
         token_filled = token.zfill(padding)
     else:
+        # For example: 00.123
         seconds, nanoseconds = token.split(".")
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/parsing.pyx:1024]: is there a risk that this raises if there is a "." but not the regex?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/parsing.pyx:1024]: yup, you'd get
```python
In [6]: token = '.'

In [7]: seconds, nanoseconds = token.split(".")

In [8]: seconds = f"{int(seconds):02d}"
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [8], line 1
----> 1 seconds = f"{int(seconds):02d}"

ValueError: invalid literal for int() with base 10: ''
```

which is what happens in the string in the linked issue (`'27.03.2003 14:55:00.000'`)

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/parsing.pyx:1024]: yeah, if the padding was `2` (say) then `'.'.zfill(2)` would give `'0.'`, and in

https://github.com/pandas-dev/pandas/blob/f7979be99585030a17f4d4a41f7447f11a9e1660/pandas/_libs/tslibs/parsing.pyx#L945-L946

it wouldn't match `parsed_formatted`

**Comment by MarcoGorelli** [comment]
merging as there's been an approval, and this is quite minor, and it unblocks another PR - if there's any other concerns / feedback I can address them in a follow-up

**Pull request merged**